libs/parallelism/algorithms/include/hpx/parallel/algorithms/swap_ranges.hpp

```diff
@@ -117,17 +94,199 @@ namespace hpx { namespace parallel { inline namespace v1 {
+
+#pragma once
+
+#include <hpx/config.hpp>
+#include <hpx/iterator_support/traits/is_iterator.hpp>
+
+#include <hpx/executors/execution_policy.hpp>
+#include <hpx/parallel/algorithms/detail/dispatch.hpp>
+#include <hpx/parallel/algorithms/detail/distance.hpp>
+#include <hpx/parallel/algorithms/for_each.hpp>
+#include <hpx/parallel/util/detail/algorithm_result.hpp>
+#include <hpx/parallel/util/projection_identity.hpp>
+#include <hpx/parallel/util/result_types.hpp>
+#include <hpx/parallel/util/zip_iterator.hpp>
+
+#include <algorithm>
+#include <iterator>
+#include <type_traits>
+#include <utility>
+
+namespace hpx { namespace parallel { inline namespace v1 {
+    template <typename Iter1, typename Iter2>
+    using swap_ranges_result = hpx::parallel::util::in_in_result<Iter1, Iter2>;
+
+    ///////////////////////////////////////////////////////////////////////////
+    // swap ranges
+    namespace detail {
+        template <typename ExPolicy, typename FwdIter1, typename FwdIter2,
+            typename Size>
+        typename util::detail::algorithm_result<ExPolicy,
+            swap_ranges_result<FwdIter1, FwdIter2>>::type
+        parallel_swap_ranges(
+            ExPolicy&& policy, FwdIter1 first1, FwdIter2 first2, Size n)
+        {
+            using zip_iterator = hpx::util::zip_iterator<FwdIter1, FwdIter2>;
+            using reference = typename zip_iterator::reference;
+
+            return get_iter_in_in_result(for_each_n<zip_iterator>().call(
+                std::forward<ExPolicy>(policy),
+                hpx::util::make_zip_iterator(first1, first2), n,
+                [](reference t) -> void {
+                    using hpx::get;
+                    std::swap(get<0>(t), get<1>(t));
+                },
+                util::projection_identity()));
+        }
+
+        template <typename IterPair>
+        struct swap_ranges
+          : public detail::algorithm<swap_ranges<IterPair>, IterPair>
+        {
+            swap_ranges()
+              : swap_ranges::algorithm("swap_ranges")
+            {
+            }
+
+            template <typename ExPolicy, typename FwdIter1, typename Sent,
+                typename FwdIter2>
+            static FwdIter2 sequential(
+                ExPolicy, FwdIter1 first1, Sent last1, FwdIter2 first2)
+            {
+                while (first1 != last1)
+                {
+                    std::swap(*first1, *first2);
+                    first1++;
+                    first2++;
+                }
+                return first2;
+            }
+
+            template <typename ExPolicy, typename FwdIter1, typename Sent1,
+                typename FwdIter2, typename Sent2>
+            static swap_ranges_result<FwdIter1, FwdIter2> sequential(ExPolicy,
+                FwdIter1 first1, Sent1 last1, FwdIter2 first2, Sent2 last2)
+            {
+                while (first1 != last1 && first2 != last2)
+                {
+                    std::swap(*first1, *first2);
```

Comment:
Shouldn't this use `std::iter_swap` instead?

Reply:
Yep, I've used the implementation of iter_swap, probably better to use iter_swap itself. Ok I'll make the change, thanks!

Reply:
To be even more precise, we actually should use `std::ranges::iter_swap`, but that's available with C++20 only. The benefit would be that it properly handles iterators that return proxy types (`std::iter_swap` does not). I have created https://github.com/STEllAR-GROUP/hpx/issues/5456 for us to remember touching all algorithms, if needed.

Reply:
This should still use `iter_swap`.